trio/_core/_multierror.py

```diff
@@ -210,41 +230,72 @@ def __new__(cls, exceptions, *, _collapse=True):
             # If this lone object happens to itself be a MultiError, then
             # Python will implicitly call our __init__ on it again.  See
             # special handling in __init__.
-            return exceptions[0]
+            single = exceptions[0]
+            assert isinstance(single, cls)
+            return single
         else:
             # The base class __new__() implicitly invokes our __init__, which
             # is what we want.
             #
             # In an earlier version of the code, we didn't define __init__ and
             # simply set the `exceptions` attribute directly on the new object.
             # However, linters expect attributes to be initialized in __init__.
+            from_class: type[Self] | type[NonBaseMultiError] = cls
             if all(isinstance(exc, Exception) for exc in exceptions):
-                cls = NonBaseMultiError
-
-            return super().__new__(cls, "multiple tasks failed", exceptions)
-
-    def __reduce__(self):
+                from_class = NonBaseMultiError
+
+            # Mypy is really mad about the following line:
+            # Ignoring arg-type: 'Argument 3 to "__new__" of "BaseExceptionGroup" has incompatible type "list[BaseException]"; expected "Sequence[_BaseExceptionT_co]"'
+            # We have checked that exceptions is indeed a list of BaseException objects, this is fine.
+            # Ignoring type-var: 'Value of type variable "Self" of "__new__" of "BaseExceptionGroup" cannot be "object"'
+            # Not sure how mypy is getting 'object', this is also fine.
+            new_obj = super().__new__(from_class, "multiple tasks failed", exceptions)  # type: ignore[arg-type,type-var]
+            assert isinstance(new_obj, (cls, NonBaseMultiError))
+            return new_obj
+
+    def __reduce__(
+        self,
+    ) -> tuple[
+        Callable[
+            [type[Self], Iterable[BaseException], DefaultNamedArg(bool, "_collapse")],
+            NonBaseMultiError | MultiError | Self,
+        ],
+        tuple[type[Self], list[BaseException]],
+        dict[str, bool],
+    ]:
```

Comment:
`mypy_extensions` isn't portable, only mypy understands it. Better to use a `Protocol` with `__call__`. But since this is only being used here, I'd suggest just typing the return type as `tuple[object, ...]`. `__reduce__`'s return is basically private, as long as it produces the right thing (up to tests to confirm), it doesn't matter what it is.

Reply:
Changed in [e0f7a9c](https://github.com/python-trio/trio/pull/2742/commits/e0f7a9cac32647c303af28f101600ade0331a45e)